### PR TITLE
idlharness support for shadowrealms

### PR DESCRIPTION
**Pull request by Ms2ger**
CC @leobalter

### CODE REVIEW

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:679]: Why not use `globalThis` here? The `self` property is set below, but why?

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3419]: Per above, is this needed?

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3421]: It looks like this can end up evaluating the scripts out of order, since the `fetch_text(url)` promises can resolve in any order. That wouldn't work, right?

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3453]: Will this run in addition to idl_test setup? If so, can they be squashed into one, so there's just one setup test?

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3426]: Why not just paste in `idls`? It will be a valid value in JS.

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3443]: Can we avoid this extra test and have just `idl_test setup`?

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3432]: Yes, if we're doing it differently in `idl_test`, please match that.

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3444]: Can values not be returned directly? Are only primitive values supported, or something?

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3449]: Can you expand on what this is? It feels similar to `fetch_tests_from_worker`, running the tests in one context but showing the results in a window context?

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3466]: Looks like this wasn't changed or really moved, but just appears to be moved due to the above addition.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js:641]: There's an initial list of spec PRs here: https://github.com/tc39/proposal-shadowrealm/issues/331. We haven't seen a use case for exposing things in ShadowRealms but not on some other globals, because those other globals could then create a ShadowRealm and use the API from there.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js:679]: idlharness predates `globalThis`, so it uses `self` everywhere. We could probably use `globalThis` everywhere now, but I didn't see a reason to start using it just here.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js:3423]: At least initially, I'd add new tests that call it explicitly, like in #31852. Maybe once ShadowRealms are widely deployed, we could do it automatically from `idl_test`.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js]: Bugger. Yeah, this is broken.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js]: So it is - left over from some older code, probably

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js:3460]: Without it, testharness.js will think we're done after the first idlharness-generated subtest. I'm not actually quite sure why, but I recall that code being fairly brittle.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js:3461]: Yeah, exactly.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js:3468]: Yes, with the difference that `fetch_tests_from_worker` and all the remote context code assume MessagePorts and asyncness, which would get in the way here.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js:3485]: I just split out the `fetch_text` function, yeah

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js:3472]: For now, I think they won't be used together.

**Review by Ms2ger — COMMENTED** [review]
Thanks for the review!

File: resources/idlharness.js
Comment (foolip) [inline resources/idlharness.js:3427]: I'm not a fan of the layering here, where idlharness.js loads itself with dependencies into that realm. And the fetching of IDL happens in the outer idlharness.js but the inner one runs the tests.

What will other tests using `ShadowRealm` look like? Will .any.js support `ShadowRealm` as well?

Can you explore having idlharness.js only once, and doing the fetching inside the `ShadowRealm`?

**Review by foolip — COMMENTED** [review]
A though about the structure of this.

But can we defer this until this is closer to landing and being implemented? I presume it's not possible to test this code at all right now.

File: resources/idlharness.js
Comment (Ms2ger) [reply resources/idlharness.js]:
```
<Ms2ger> foolip: so, the thing is that code inside a ShadowRealm doesn't get to fetch anything - only import JS modules
<Ms2ger> foolip: alternatively, I can extract my new code from idlharness.js into its own file, and just duplicate the fetch_spec thing
<foolip> Hmm, I thought I saw fetch() in https://github.com/tc39/proposal-shadowrealm/issues/331, but I didn't.
<Ms2ger> That would have made my life much easier, yes :)
<foolip> So, I can't spend more time in the shadows today, but am willing to be pinged about this again in the future. I always have a lot of time in the future.
<Ms2ger> Heh
<Ms2ger> foolip: in terms of future - this is blocking merging the spec PRs, at least for whatwg. I don't mind holding off a bit, but not long
<foolip> I hope this isn't the only test for ShadowRealm in WPT?
<Ms2ger> foolip: wpt only has the ones for Web APIs; the rest is in https://github.com/tc39/test262/tree/main/test/built-ins/ShadowRealm
```

I pushed the alternative approach; please take a look if you have a moment.

I don't have clear insight in what other tests will look like - at this point I'm not convinced we'll end up writing so many tests that more infrastructure will be needed. I think we should look into that when we have some more tests to extract use cases from.

File: resources/idlharness.js
Comment (rwaldron) [reply resources/idlharness.js]: Callables (functions or callable Proxy objects), Primitives and Symbols. Note that a callable itself is also limited to accepting only values of those types for arguments and will return only values for those types.

**Comment by rwaldron** [comment]
@Ms2ger great stuff here, thanks!

**Comment by foolip** [comment]
Alright, I like this second variant much better. Feel free to ping me for proper review next week.

File: resources/idlharness-shadowrealm.js
Comment (Ms2ger) [inline resources/idlharness-shadowrealm.js:20]: Now that this is in a separate file, I guess this might as well use async/await. @foolip, do you have a preference?

Comment (foolip) [reply resources/idlharness-shadowrealm.js:20]: If you want to rewrite it to async/await I wouldn't mind, but if you can't be bothered this is OK too.

File: resources/idlharness-shadowrealm.js
Comment (foolip) [inline resources/idlharness-shadowrealm.js:37]: I guess this is because it's not in a window context, so it's similar to workers needing an explicit `done()`? So can we use that instead and avoid the wrapping test?

File: resources/idlharness-shadowrealm.js
Comment (foolip) [inline resources/idlharness-shadowrealm.js:20]: Can we use `promise_setup` instead of a test to avoid the extra test?

File: resources/idlharness-shadowrealm.js
Comment (foolip) [inline resources/idlharness-shadowrealm.js:56]: Yes, this is rather cryptic. Is this how results from a worker context are transferred too? Is the use of `async_test` just to avoid something obscure about the timing when testharness.js considers all results complete?

File: resources/idlharness-shadowrealm.js
Comment (foolip) [inline resources/idlharness-shadowrealm.js:65]: There's no `IdlHarnessError` here.

File: resources/idlharness-shadowrealm.js
Comment (foolip) [inline resources/idlharness-shadowrealm.js:76]: We don't actually use the `spec` property, so the `fetch_spec` wrapper is kind of silly. Can you just inline the URL construction into a `fetch_text` call instead?

File: resources/idlharness-shadowrealm.js
Comment (foolip) [inline resources/idlharness-shadowrealm.js:22]: Can you add a TODO here about https://github.com/web-platform-tests/wpt/pull/31851/files#r761766428?

File: resources/idlharness-shadowrealm.js
Comment (Ms2ger) [reply resources/idlharness-shadowrealm.js:56]: This was basically a suggestion from James. I'm not sure if `test()` worked - in any case, after `promise_setup`, only `promise_test` is supported, so that's what it uses now. The goal is to move this to th.js Soon(TM), probably in early January, but I'd like to land this in its current hack state before that.

File: resources/idlharness-shadowrealm.js
Comment (Ms2ger) [reply resources/idlharness-shadowrealm.js:37]: I have looked briefly, but this part of th.js is... interesting. I'd be willing (though not very excited) to look into this more, but I'd prefer to land first so I can unblock my spec work.

**Review by Ms2ger — COMMENTED** [review]
Thanks!

File: resources/idlharness-shadowrealm.js
Comment (foolip) [reply resources/idlharness-shadowrealm.js:68]: Oh, delightful. I'm going to go 👀 and this TODO will have to do.

File: resources/idlharness-shadowrealm.js
Comment (foolip) [reply resources/idlharness-shadowrealm.js:49]: I feel like this is adding another layer of dependencies on weird behavior or testharness.js that none of us really understand, but I also can't be bothered to look into this now, so I can't block you :)

**Review by foolip — APPROVED** [review]
Hope it works :)

**Pull request merged**
